### PR TITLE
[release tool] refactor pinnedversion

### PR DESCRIPTION
**Pull request by radTuti**
## Description

This refactors the pinnedversion package to allow for different pinnedversion per product.

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
TBD
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: release/pkg/manager/calico/manager.go [hunk 1]
```diff
@@ -203,6 +202,7 @@ type CalicoManager struct {
 	// image scanning configuration.
 	imageScanning       bool
 	imageScanningConfig imagescanner.Config
+	imageComponents map[string]registry.Component
```
Comment (caseydavenport) [inline release/pkg/manager/calico/manager.go:205]: formatting looks off here - goimports?

File: release/internal/pinnedversion/pinnedversion.go [hunk 2]
```diff
@@ -47,167 +48,222 @@ var noImageComponents = []string{
...
 
-	// Registry is the registry for Tigera operator
+type OperatorConfig struct {
+	Dir      string
+	Branch   string
 	Registry string
-
-	// Image is the image for Tigera operator
-	Image string
+	Image    string
```
Comment (caseydavenport) [inline release/internal/pinnedversion/pinnedversion.go:60]: I think it was good to remove the other comments. Image is a tricky one - might be worth a comment explaining the expected format.

e.g., does this take `tigera/operator`? Or does it take `quay.io/tigera/operator`? Or does it take `tigrera/opreator:tag`? etc.

File: release/internal/pinnedversion/pinnedversion.go [hunk 3]
```diff
@@ -47,167 +48,222 @@ var noImageComponents = []string{
...
+	Components     map[string]registry.Component `yaml:"components"`
 }
 
-// PinnedVersionData represents the data needed to generate the pinned version file from the template.
-type PinnedVersionData struct {
-	// ReleaseName is the name of the release.
-	ReleaseName string
+func NewCalicoVersionData(productVersion, operatorVersion string) *CalicoPinnedVersionData {
+	return &CalicoPinnedVersionData{
+		calico:   version.New(productVersion),
```
Comment (caseydavenport) [inline release/internal/pinnedversion/pinnedversion.go:90]: It seems a little odd to me to parse these into `Version` structs, and then throughout the `CalicoPinnedVersionData` implementation we exclusively call `FormattedString` to convert back into a string. Is there a reason?

File: release/internal/pinnedversion/pinnedversion.go [hunk 4]
```diff
@@ -47,167 +48,222 @@ var noImageComponents = []string{
...
+	return fmt.Sprintf("https://%s.%s", d.ReleaseName, d.BaseDomain)
+}
 
 func pinnedVersionFilePath(outputDir string) string {
 	return filepath.Join(outputDir, pinnedVersionFileName)
 }
 
-func operatorComponentsFilePath(outputDir string) string {
-	return filepath.Join(outputDir, operatorComponentsFileName)
+type CalicoPinnedVersions struct {
```
Comment (caseydavenport) [inline release/internal/pinnedversion/pinnedversion.go:139]: Confusing to have structs named `CalicoPinnedVersions` and `CalicoPinnedVersionData` in the same file without any comments explaining the separation - what is the intended scope of these two types?

File: release/internal/pinnedversion/pinnedversion.go (hunk 3 above)
Comment (caseydavenport) [reply release/internal/pinnedversion/pinnedversion.go:90]: I'm actually not sure what the purpose of this structure - `CalicoPinnedVersionData` - is.

I originally thought it was going to implement the `PinnedVersions` interface, but looks like that is actually done by `CalicoPinnedVersions` instead.

This struct seems to effectively take two strings and then provide methods to access those strings, which seems like it might be overcomplicated rather than just passing the strings around where needed. Unless there's an interface that this is intended to satsify with an alternative implementation?

Basically, I think the code is missing comments justifying why this construct should exist.

**Review by caseydavenport — CHANGES_REQUESTED** [review]
A few minor comments / questions

File: release/pkg/manager/calico/manager.go (hunk 1 above)
Comment (radTuti) [reply release/pkg/manager/calico/manager.go:205]: i missed adding it after i ran `make fix-changed`

File: release/internal/pinnedversion/pinnedversion.go (hunk 3 above)
Comment (radTuti) [reply release/internal/pinnedversion/pinnedversion.go:90]: I have moved this struct to `version` package. It implements `version.Data` interface
